### PR TITLE
automatic layers

### PR DESCRIPTION
**Pull request by kapilt**
closes #1063  related #1127 

Automatic construction of layer for a chalice app (by stage) and reused across all an app's lambda functions. Its configurable and enabled by default. Old layers are automatically removed on deploy and delete.

### CODE REVIEW

**Comment by kapilt** [comment]
ready for review/feedback. the test failure on 2.7 seems to be a travis anomaly wrt to docutils.

**Comment by kapilt** [comment]
let's go ahead and add config for it default on, and auto delete.. ordering delete to happen after publish and config on existing lambdas.

File: chalice/package.py [hunk 1]
```diff
@@ -136,10 +151,15 @@ def _generate_lambdafunction(self, resource, template):
             }
             lambdafunction_definition['Properties'].update(
                 reserved_concurrency_config)
-        if resource.layers:
+
+        layers = resource.layers or []  # type: List[Any]
+        if self._chalice_layer:
+            layers.append({'Ref': self._chalice_layer})
```
Comment (kapilt) [inline chalice/package.py]: this needs to be inserted at the start, we want user explicit layers to have precedence.

File: chalice/deploy/sweeper.py [hunk 2]
```diff
@@ -78,79 +170,8 @@ def _plan_deletion(self,
         # type: (...) -> None
         for name in remaining:
             resource_values = deployed.resource_values(name)
-            if resource_values['resource_type'] == 'lambda_function':
```
Comment (kapilt) [inline chalice/deploy/sweeper.py]: Fwiw, This for/if loop refactoring was triggered by hitting flake8  check on cyclomatic complexity limits when adding another conditional.

**Comment by jamesls** [comment]
Just a heads up, I'm going to pick up this work again and try to get this merged.  I'll get it rebased on the latest master and test it out.

**Comment by jamesls** [comment]
I've just about got everything rebased and working, but in testing this out, I'm noticing a difference in behavior with layers.  For layers, the zip file is unzipped to `/opt`, which means that in order for us to get the same behavior as before with no layers, when creating the layer zip file we need to unzip to `python/lib/{python-version}/site-packages` (documented [here](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html#configuration-layers-manage)).

While that works for python files, it's not going to work for `vendor/` files.  That's because vendor files either can be python packages that you were intending to import (i.e custom built python packages), *or* they can be static assets you just want dumped into the lambda function.  That means that today if I have a file that's in `./vendor/my-picture.png`, then in my app code I can say `with open('my-picture.png')` and it'll work.  However, with layers, that file will be over in `/opt/python` somewhere.

Because of this incompatibility, I don't think we'll be able to default to using automatic layers.  This will have to be something you opt into for the time being until we can come up with a way to smooth other this difference.

**Comment by jamesls** [comment]
Closing, https://github.com/aws/chalice/pull/1487 has been merged/released.

**Pull request closed without merging**